### PR TITLE
bpo-35134: Create Include/cpython/ subdirectory

### PR DESCRIPTION
**Pull request by vstinner**
Include/*.h should be the "portable Python API", whereas
Include/cpython/*.h should be the "CPython API": CPython
implementation details.

Changes:

* Create Include/cpython/ subdirectory
* "make install" now creates $prefix/include/cpython and copy
  Include/cpython/* to $prefix/include/cpython
* Create Include/cpython/objimpl.h: move objimpl.h code
  surrounded by "#ifndef Py_LIMITED_API" to cpython/objimpl.h.
* objimpl.h now includes cpython/objimpl.h

https://bugs.python.org/issue35134

### CODE REVIEW

**Comment by vstinner** [comment]
I rebased my change on top of master to retrieve PR #10626 bugfix.

**Comment by vstinner** [comment]
@ncoghlan, @serhiy-storchaka, @pitrou, @zooba, @ericsnowcurrently: Would you mind to review this PR? I created a new Include/unstable/ subdirectory for all "#ifndef Py_LIMITED_API" APIs. See https://bugs.python.org/issue35134 for the rationale.

**Comment by vstinner** [comment]
I plan to merge this PR next week.

File: Include/unstable/objimpl.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef _Py_UNSTABLE_OBJIMPL_H
+#  error "this header file must not be included directly"
```
Comment (ncoghlan) [inline Include/unstable/objimpl.h]: I was going to quibble about the directory name, but given this style of inclusion (i.e. only implicitly via the public header file when `Py_LIMITED_API` is *not* defined), I think 'unstable' is OK.

**Review by ncoghlan — APPROVED** [review]

File: Include/unstable/objimpl.h (hunk 1 above)
Comment (vstinner) [reply Include/unstable/objimpl.h]: I mostly wrote this #error to prevent the risk of bikeshedding, LOL!

**Comment by vstinner** [comment]
@ncoghlan: You approved the PR but proposed a different directory name. Are you ok with "unstable" name? https://bugs.python.org/issue35134#msg330261

**Comment by vstinner** [comment]
Note: this PR is my second attempt, the first one was PR #10285 which wasn't properly designed.

**Comment by vstinner** [comment]
@zooba: Aha, I was looking for the code of the Windows installer.

> Tools/msi/dev/dev.wixproj is going to need its <InstallFiles Include="$(PySourcePath)include\*.h"> line changed to <InstallFiles Include="$(PySourcePath)include\**\*.h">

Right now, Include/internal/ is **not** installed (on Unix at least). I just proposed to install these headers as well: https://bugs.python.org/issue35296

Would it be possible to only update the MSI installer once this PR and PR #10665 are merged?

**Comment by vstinner** [comment]
I renamed the subdirectory from Include/unstable/ to Include/cpython/.

**Comment by zooba** [comment]
@vstinner If you're willing to keep reminding me to do it once this is merged until it's done, sure.

**Comment by vstinner** [comment]
> @vstinner If you're willing to keep reminding me to do it once this is merged until it's done, sure.

I merged my PR #10665 (make install now installs the internal API) and I udpated this PR to apply your proposed change in MSI (copy `Include/**/*.h`).

**Comment by ncoghlan** [comment]
@vstinner While I preferred "cpython" as the directory name, I was OK with "unstable", hence the original approval. I'm happy you decided to switch the name anyway, though :)

**Comment by vstinner** [comment]
:-)

**Pull request merged**